gradio/route_utils.py

```diff
@@ -583,3 +584,54 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+def get_core_url(original_url: str) -> str:
+    """
+    Returns the core url of a given url. If the url does not have a scheme, it is assumed to be http.
+    Examples:
+        get_core_url("https://www.gradio.app") -> "www.gradio.app"
+        get_core_url("localhost:7860") -> "localhost"
+        get_core_url("127.0.0.1") -> "127.0.0.1"
+    """
+    if not original_url:
+        return ""
+    if "://" not in original_url:
+        original_url = "http://" + original_url
+    return httpx.URL(original_url).host
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = request.headers.get("origin", "")
+        host_header_host = get_core_url(host_header)
+        origin_header_host = get_core_url(origin_header)
+
+        localhost_aliases = ["localhost", "127.0.0.1", "0.0.0.0", "null"]
+
+        if (
```

Comment:
While the current impl probably works for the purpose, I think it's not CORS actually.
In CORS, the server only informs the allowed origin to the browser, and the browser is actually responsible to restrict the cross-origin resource access based on the allowed origin info.
See https://stackoverflow.com/a/36959539

To be compliant these specs, the code might be like this? WDYT?
```python
accessed_host_is_localhost = host_header_host in localhost_aliases
is_preflight = request.method == "OPTIONS"
if accessed_host_is_localhost and is_preflight:
    # For non-simple requests which can causes side-effects on the server in localhost,
    # we only allow localhost to access the endpoint.
    response = fastapi.Response()
    response.headers["Access-Control-Allow-Origin"] = "http://localhost"  # TODO: This line can be improved to cover more localhost origins like "127.0.0.0"
    return response

# In other cases, all origins are allowed.
if request.method == "OPTIONS":
    response = fastapi.Response()
else:
    response = await call_next(request)

response.headers["Access-Control-Allow-Origin"] = request.headers.get("origin", "*")
return response

```

Reply:
Interesting, I see.. this makes sense to me. Let me test this out

Reply:
Ok I've made the change to let the browser enforce CORS. One difference is that I needed to set the CORS headers on all requests, not just the preflight requests.